File: ios/PetitesGouttes/ViewModels/FreezerViewModel.swift
import SwiftUI
import SwiftData

enum SortOrder: String, CaseIterable {
    case dateAsc = "Date (ancien)"
    case dateDesc = "Date (recent)"
    case volumeAsc = "Volume (croissant)"
    case volumeDesc = "Volume (decroissant)"
}

@Observable
@MainActor
final class FreezerViewModel {
    private let modelContext: ModelContext
    var activeBags: [MilkBag] = []
    var sortOrder: SortOrder = .dateAsc

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
        fetchBags()
    }

    func fetchBags() {
        let predicate = #Predicate<MilkBag> { !$0.removedFromFreezer }
        let descriptor = FetchDescriptor<MilkBag>(predicate: predicate)
        activeBags = sortedBags((try? modelContext.fetch(descriptor)) ?? [])
    }

    func removeBag(_ bag: MilkBag) {
        bag.removedFromFreezer = true
        bag.removalDate = .now
        try? modelContext.save()
        fetchBags()
    }

    func deleteBag(_ bag: MilkBag) {
        modelContext.delete(bag)
        try? modelContext.save()
        fetchBags()
    }

    func setSortOrder(_ order: SortOrder) {
        sortOrder = order
        activeBags = sortedBags(activeBags)
    }

    private func sortedBags(_ bags: [MilkBag]) -> [MilkBag] {
        switch sortOrder {
        case .dateAsc: bags.sorted { $0.pumpDate < $1.pumpDate }
        case .dateDesc: bags.sorted { $0.pumpDate > $1.pumpDate }
        case .volumeAsc: bags.sorted { $0.volumeMl < $1.volumeMl }
        case .volumeDesc: bags.sorted { $0.volumeMl > $1.volumeMl }
        }
    }
}

File: app/src/main/java/com/bnjdpn/petitesgouttes/viewmodel/FreezerViewModel.kt
package com.bnjdpn.petitesgouttes.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.bnjdpn.petitesgouttes.data.database.AppDatabase
import com.bnjdpn.petitesgouttes.data.database.MilkBagEntity
import com.bnjdpn.petitesgouttes.data.repository.MilkBagRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.launch

enum class SortOrder { DATE_ASC, DATE_DESC, VOLUME_ASC, VOLUME_DESC }

class FreezerViewModel(application: Application) : AndroidViewModel(application) {
    private val repository: MilkBagRepository

    val activeBags: Flow<List<MilkBagEntity>>

    private val _sortOrder = MutableStateFlow(SortOrder.DATE_ASC)
    val sortOrder: StateFlow<SortOrder> = _sortOrder

    init {
        val dao = AppDatabase.getInstance(application).milkBagDao()
        repository = MilkBagRepository(dao)
        activeBags = repository.activeBags
    }

    fun setSortOrder(order: SortOrder) {
        _sortOrder.value = order
    }

    fun sortBags(bags: List<MilkBagEntity>, order: SortOrder): List<MilkBagEntity> {
        return when (order) {
            SortOrder.DATE_ASC -> bags.sortedBy { it.pumpDate }
            SortOrder.DATE_DESC -> bags.sortedByDescending { it.pumpDate }
            SortOrder.VOLUME_ASC -> bags.sortedBy { it.volumeMl }
            SortOrder.VOLUME_DESC -> bags.sortedByDescending { it.volumeMl }
        }
    }

    fun removeBag(id: Long) {
        viewModelScope.launch {
            repository.removeBag(id)
        }
    }

    fun deleteBag(bag: MilkBagEntity) {
        viewModelScope.launch {
            repository.deleteBag(bag)
        }
    }
}
